index.ts

```diff
@@ -17,50 +17,47 @@ abstract class NgComponent<Props, State> {
   */
   // nb: this method is explicity exposed for unit testing
   public $onChanges(changes: {}) {
+    console.info('TODOCK onChanges')
     const oldProps = mapValues<{}, Props>(changes, 'previousValue')
     const newProps = mapValues<{}, Props>(changes, 'currentValue')
 
     if (!this.__isFirstRender) {
       this.componentWillReceiveProps(newProps)
     }
 
-    if (this.__isFirstRender || this.didPropsChange(newProps, oldProps)) {
+    // store the new props
+    this.props = assign({}, this.props, newProps)
 
-      // compute whether we should render or not
+    if (this.__isFirstRender) {
+      this.render()
+      this.__isFirstRender = false
+
+    } else if (this.didPropsChange(newProps, oldProps)) {
       const shouldUpdate = this.shouldComponentUpdate(
         assign({}, this.props, newProps),
         assign({}, this.props, oldProps)
       )
 
-      // store the new props
-      this.props = assign({}, this.props, newProps)
-
       if (shouldUpdate) {
-
-        if (!this.__isFirstRender) {
-          this.componentWillUpdate(this.props, this.state)
-        }
-
-        this.render(this.props, this.state)
-
-        if (!this.__isFirstRender) {
-          this.componentDidUpdate(this.props, this.state)
-        }
-
-        this.__isFirstRender = false
+        this.componentWillUpdate(this.props, this.state)
+        this.render()
+        this.componentDidUpdate(this.props, this.state)
       }
     }
   }
 
   $onInit() {
+    console.info('TODOCK componentWillMount')
     this.componentWillMount()
   }
 
   $postLink() {
+    console.info('TODOCK componentDidMount')
     this.componentDidMount()
   }
 
   $onDestroy() {
+    console.info('TODOCK componentWillUnmount')
```

Comment:
trying to fix a problem

![image](https://cloud.githubusercontent.com/assets/681288/23974609/570d6bc2-0999-11e7-805a-d6078f48d8ea.png)

componentWillMount & componentDidMount gets called after onChange event (e. where render gets called)... seriously out of order